docs/adr/adr-007-monitoring-and-alerting.md

```diff
@@ -0,0 +1,267 @@
+unactionable alerts and reviewing the GOV.UK incidents for Email Alert API
+helped us to identify patterns where incidents reflected ambiguity over whether
+the system was broken or not. We felt there was room for improvement.
+
+[monitor-time-frame]: https://github.com/alphagov/email-alert-api/blob/c70fcd7e5d64393074a3eebd83b3467fd8cc03b3/app/workers/digest_run_worker.rb#L6-L9
+[monitor-workload]: https://github.com/alphagov/govuk-puppet/blob/fac4b5163f10736bb19aa51a8eb7bc276dd2cb40/modules/govuk/manifests/apps/email_alert_api/sidekiq_queue_check.pp#L33-L44
+[email-alert-monitoring]: https://github.com/alphagov/email-alert-monitoring
+[Travel Advice Publisher]: https://github.com/alphagov/travel-advice-publisher
+
+## Decision
+
+We decided that the system should raise a critical alert
+[if and only if](https://en.wikipedia.org/wiki/If_and_only_if) the email alert
+system is broken. The majority of these alerts should contact engineers
+out-of-hours.
+
+Unpacking this, we consider the email alert system to be comprised of 3
+applications: Email Alert API,
+[Email Alert Frontend](https://github.com/alphagov/email-alert-frontend) and
+[Email Alert Service](https://github.com/alphagov/email-alert-service); this
+excludes the monitoring of whether GOV.UK publishing is working and whether
+Notify is working as they are independently monitored. The use of
+"if and only if" reflects a condition we've formed that the system will not
+raise critical alerts unless we're highly confident it is broken, whereas
+warnings are appropriate for other situations. We decided that for the email
+alert system to be "broken" it represents a failing that will lead to a user
+noticeable issue unless an engineer intervenes. Since these alerts should
+represent a broken system requiring intervention, it should be suitable to
+contact engineers out-of-hours for them as otherwise users will be
+experiencing issues.
+
+We decided that we will re-approach our alerting with this mindset and look to
+replace alerts that are vulnerable to reflecting a busy status, or which
+extend beyond the boundaries of the system.
+
+## Status
+
+Pending
+
+## Consequences
+
+We will be changing the approaching to monitoring and alerting across the email
+alert system of applications in order to meet this new alerting criteria.
+
+### Monitoring Email Alert API
+
+#### Sidekiq workers
+
+We will configure critical alerts, that contact engineers out-of-hours, if
+individual Sidekiq jobs fail multiple times for the following workers:
+
+- [DailyDigestInitiatorWorker][]
+- [DigestEmailGenerationWorker][]
+- [ProcessContentChangeWorker][]
+- [ProcessMessageWorker][]
+- [SendEmailWorker][]
+- [WeeklyDigestInitiatorWorker][]
+
+These workers will use the database to store the number of times we have
+attempted to perform an action. An alert will be raised if the quantity of
+failed attempts reaches a threshold (the suggestion is 5). To ensure that these
```

Comment:
Could you elaborate on why 5 was chosen? 

Does this number also need to be adjusted per worker? 5 failed attempts trying to send a single email seems minor, 5 failed attempts trying to process a daily digest seems major.

Reply:
I've added a brief comment to justify 5. The same would be applied to all the workers. We worked out that with intermittent network problems we'd be extremely unlucky to have the same email 5 times. All the other workers don't make external requests to services so should be very unlikely to fail.

There's an aspect from the original document of seeing if we got any SendEmailWorker false positives due to intermittent failures. Ideas on approaching it could be automatically retrying within the worker on SendEmailWorker failures or setting that worker to be more lenient. Hopefully this won't be needed though.

I'm not too sure how to add this into the ADR without it going off on a tangent... 🤔

Reply:
Your brief comment does help but could we add something along the lines of your first paragraph as I think it's helpful to highlight the decision up on why 5 was chosen and applied to all workers equally?

Reply:
Sure, I've updated it some more.

Reply:
Much clearer, thanks!